aten/src/ATen/core/ivalue.cpp

```diff
@@ -470,6 +470,69 @@ std::shared_ptr<ClassType> ivalue::Object::type() const {
   return type_.type_->expect<ClassType>();
 }
 
+IValue IValue::deepcopy() const {
+  IValue::HashAliasedIValueMap memo;
+  return deepcopy(memo);
+}
+
+IValue IValue::deepcopy(
+    IValue::HashAliasedIValueMap& memo) const {
+  if (memo.count(*this)) {
+    return memo.at(*this);
+  }
+  IValue copy;
+  switch(tag) {
+    case IValue::Tag::Tensor:
+      copy = IValue(toTensor().clone());
+      break;
+    case IValue::Tag::Tuple: {
+      std::vector<IValue> copied_tuple;
+      for (const auto& e : toTuple()->elements()) {
+        copied_tuple.push_back(e.deepcopy(memo));
+      }
+      copy = IValue(ivalue::Tuple::create(copied_tuple));
+    }
+      break;
+    case IValue::Tag::GenericList: {
+      auto list = toList();
+      auto copied_list = c10::impl::GenericList(list.elementType());
+      for (IValue v : list) {
+        copied_list.push_back(v.deepcopy(memo));
+      }
+      copy = IValue(copied_list);
+    }
+      break;
+    case IValue::Tag::GenericDict: {
+      auto dict = toGenericDict();
+      auto copied_dict = c10::impl::GenericDict(dict.keyType(), dict.valueType());
+      for (const auto& entry : dict) {
+        copied_dict.insert(entry.key().deepcopy(memo), entry.value().deepcopy(memo));
+      }
+      copy = IValue(copied_dict);
+    }
+      break;
+    case IValue::Tag::Object: {
+      copy = IValue(toObject()->deepcopy(memo));
+      break;
+    case IValue::Tag::String:
+    case IValue::Tag::None:
+    case IValue::Tag::Double:
+    case IValue::Tag::Int:
+    case IValue::Tag::Bool:
+    case IValue::Tag::Device:
+    case IValue::Tag::Uninitialized:
+      copy = *this;
+      break;
+    default:
+      AT_ERROR("Can't deepcopy IValue with tag: ", tagKind());
+    }
+  }
+  if (!isAliasOf(copy)) {
+    memo[*this] = copy;
```

Comment:
Using an Interface type today, you can actually get an object to contain itself. If this memo field isn't written two _before_ making recursive calls to deep copy, then this code is going to stack overflow. I'm not sure how much we should try to fix this now. On the one hand, it is very rare such that no models today do this. On the other hand, sometime in the future we are going to expand the object system and this code will start crashing.

Reply:
I see, should we move this to L484 and check the address of IValue then?
do we have the concept of the address or ID of an IValue?

Reply:
I'll add a comment for now, we can have a follow up PR later.